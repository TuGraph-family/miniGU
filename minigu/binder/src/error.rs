use std::fmt::{Debug, Display, Formatter};
use std::marker::PhantomData;
use std::ops::Range;
use std::sync::Arc;
use std::result;
use gql_parser::error::TokenError;
use crate::{impl_binder_already_exist_error_display, impl_binder_not_exist_error_display};


#[cfg(feature = "miette")]
use miette::Diagnostic;
use thiserror::Error;
pub type Result<T> = result::Result<T, Error>;

#[derive(Error, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Clone)]
pub enum Error {
    #[error("schema not exists {0:?}")]
    SchemaNotExists(String),
    #[error("schema already exists {0:?}")]
    SchemaAlreadyExists(String),
    #[error("graph type not exists {0:?}")]
    GraphTypeNotExists(String),
    #[error("graph type already exists {0:?}")]
    GraphTypeAlreadyExists(String),
    #[error("graph not exists {0:?}")]
    GraphNotExists(String),
    #[error("graph already exists {0:?}")]
    GraphAlreadyExists(String),
    #[error("procedure {0} not exists")]
    ProcedureNotExists(String),
    #[error("node type not exists {0:?}")]
    NodeTypeNotExists(String),
    #[error("edge type not exists {0:?}")]
    EdgeTypeNotExists(String),
    #[error("invalid field {0:?}")]
    InvalidField(String),
    #[error("duplicate field name {0:?}")]
    DuplicateField(String),
    #[error("not support operation {0:?}")]
    NotSupported(String),
    #[error("tmp error")]
    ErrorCur
}

// #[derive(Debug, Clone, PartialEq, Eq)]
// #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
// pub struct BinderError<T> {
//     input: Arc<str>,
//     span: Range<usize>,
//     position: (usize, usize),
//     #[cfg_attr(feature = "serde", serde(skip))]
//     _marker: PhantomData<T>,
// }

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NotSupportError {
    pub operation_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InvalidFieldError {
    pub field_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InvalidKeyError {
    pub key_name: String,
}



// impl<T> BinderError<T> {
//     pub fn input(&self) -> &Arc<str> {
//         &self.input
//     }
// 
//     pub fn span(&self) -> &Range<usize> {
//         &self.span
//     }
// 
//     pub fn position(&self) -> (usize, usize) {
//         self.position
//     }
// }

impl NotSupportError {
    pub fn operation_name(&self) -> &String {
        &self.operation_name
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SchemaNotExistsError {
    pub schema_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SchemaAlreadyExistsError {
    pub schema_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GraphTypeNotExistsError {
    pub type_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GraphTypeAlreadyExists {
    pub type_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GraphNotExistsError {
    pub graph_name: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GraphAlreadyExists {
    pub graph_name: String,
}

// impl_binder_not_exist_error_display!(
//     (SchemaNotExists, "schema"),
//     (GraphNotExists, "graph"),
//     (GraphTypeNotExists, "graph type"),
// );
//
// impl_binder_already_exist_error_display!(
//     (SchemaAlreadyExists, "schema"),
//     (GraphAlreadyExists, "graph"),
//     (GraphTypeAlreadyExists, "graph type"),
// );

impl Display for NotSupportError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let operation = &self.operation_name();
        write!(f,
            "{operation} is not support yet"
        )
    }
}
// 
// impl<T> core::error::Error for BinderError<T> where BinderError<T>: Debug + Display {}
impl core::error::Error for NotSupportError {}
// 
// #[cfg(feature = "miette")]
// impl<T> Diagnostic for BinderError<T> where BinderError<T>: core::error::Error {
//     fn source_code(&self) -> Option<&dyn miette::SourceCode> {
//         Some(&self.input)
//     }
// 
//     fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
//         Some(Box::new(core::iter::once(
//             miette::LabeledSpan::new_with_span(Some("here".into()), self.span.clone()),
//         )))
//     }
// }

#[cfg(feature = "miette")]
impl Diagnostic for NotSupportError {
    fn source_code(&self) -> Option<&dyn miette::SourceCode> {
        Some(&self.operation_name)
    }
}

