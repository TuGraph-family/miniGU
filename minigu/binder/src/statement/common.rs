use serde::Serialize;
use crate::types::Ident;

pub type Yield = Vec<YieldItem>;

#[derive(Debug, Serialize)]
pub struct YieldItem {
    pub name: Ident,
    pub alias: Option<Ident>,
}