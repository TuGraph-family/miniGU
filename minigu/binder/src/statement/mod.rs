pub mod catalog;
mod common;
mod object_ref;
mod procedure;
mod type_element;
