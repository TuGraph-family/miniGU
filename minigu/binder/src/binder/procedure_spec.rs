use std::collections::HashMap;

use gql_parser::ast::{Procedure, Statement};
use itertools::Itertools;
use smol_str::SmolStr;

use super::Binder;
use crate::error::{BindResult, not_implemented};
use crate::procedure::procedure_spec::{BoundProcedure, BoundStatement};

impl Binder {
    pub fn bind_procedure(&mut self, procedure: &Procedure) -> BindResult<BoundProcedure> {
        if let Some(schema) = &procedure.at {
            let schema = self.bind_schema_ref(schema.value())?;
            self.current_schema = Some(schema);
        }
        if !procedure.binding_variable_defs.is_empty() {
            return not_implemented("binding variable definitions".to_string(), None);
        }
        let statement = self.bind_statement(procedure.statement.value())?;
        if !procedure.next_statements.is_empty() {
            return not_implemented("next statement".to_string(), None);
        }
        Ok(BoundProcedure {
            statement,
            next_statements: Vec::new(),
        })
    }

    pub fn bind_statement(&mut self, statement: &Statement) -> BindResult<BoundStatement> {
        match statement {
            Statement::Catalog(statements) => statements
                .iter()
                .map(|s| self.bind_catalog_modifying_statement(s.value()))
                .try_collect()
                .map(BoundStatement::Catalog),
            Statement::Query(statement) => {
                todo!()
            },
            Statement::Data(_) => {
                not_implemented("data-modifying statement".to_string(), None)
            }
        }
    }
}


