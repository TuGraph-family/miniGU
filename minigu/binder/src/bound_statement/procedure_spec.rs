use macro_rules_attribute::apply;
use serde::Serialize;
use gql_parser::ast::{BindingVariableDefBlock, Yield};
use gql_parser::span::OptSpanned;
use crate::binder::bound_statement::data::BoundLinearDataModifyingStatement;
use crate::binder::bound_statement::query::BoundCompositeQueryStatement;
use crate::bound_statement::catalog::LinearBoundCatalogModifyingStatement;
use crate::catalog_ref::SchemaCatalog;
use crate::macros::base;

#[derive(Debug, Serialize)]
pub struct BoundProcedure {
    pub at: Option<SchemaCatalog>,
    pub binding_variable_def: BindingVariableDefBlock,
    pub statement: BoundStatement,
    pub next_statement: Vec<BoundNextStatement>,
}


#[apply(base)]
pub enum BoundStatement {
    Catalog(LinearBoundCatalogModifyingStatement),
    Query(BoundCompositeQueryStatement),
    Data(BoundLinearDataModifyingStatement),
}


#[apply(base)]
pub struct BoundNextStatement {
    pub yield_clause: OptSpanned<Yield>,
    pub statement: BoundStatement,
}
