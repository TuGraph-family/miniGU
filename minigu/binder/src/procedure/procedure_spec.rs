use serde::Serialize;

use super::catalog::BoundCatalogModifyingStatement;

#[derive(Debug, Serialize)]
pub struct BoundProcedure {
    pub statement: BoundStatement,
    pub next_statements: Vec<BoundNextStatement>,
}

#[derive(Debug, Serialize)]
pub enum BoundStatement {
    Catalog(Vec<BoundCatalogModifyingStatement>),
    // Query(BoundCompositeQueryStatement),
    // Data(BoundLinearDataModifyingStatement),
}

#[derive(Debug, Serialize)]
pub struct BoundNextStatement {
    pub yield_column_indices: Vec<usize>,
    pub statement: BoundStatement,
}
