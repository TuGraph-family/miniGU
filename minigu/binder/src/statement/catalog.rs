use serde::Serialize;

use crate::object_ref::{GraphRef, SchemaRef};
use crate::statement::object_ref::{BoundGraphType, CanonicalSchemaPath};
use crate::statement::procedure::BoundCallProcedureStatement;
use crate::types::Ident;

pub type LinearBoundCatalogModifyingStatement = Vec<BoundCatalogModifyingStatement>;
#[derive(Debug, Serialize)]
pub enum BoundCatalogModifyingStatement {
    Call(BoundCallProcedureStatement),
    CreateSchema(BoundCreateSchemaStatement),
    DropSchema(BoundDropSchemaStatement),
    CreateGraph(BoundCreateGraphStatement),
    DropGraph(BoundDropGraphStatement),
    CreateGraphType(BoundCreateGraphTypeStatement),
    DropGraphType(BoundDropGraphTypeStatement),
}

#[derive(Debug, Serialize)]
pub struct BoundCreateSchemaStatement {
    pub schema_path: CanonicalSchemaPath,
    pub if_not_exists: bool,
}

#[derive(Debug, Serialize)]
pub struct BoundDropSchemaStatement {
    pub schema_path: CanonicalSchemaPath,
    pub if_exists: bool,
}

#[derive(Debug, Serialize)]
pub struct BoundCreateGraphStatement {
    pub schema: SchemaRef,
    pub name: Ident,
    pub kind: CreateKind,
    pub type_ref: BoundGraphType,
    pub source: GraphRef,
}

#[derive(Debug, Serialize)]
pub enum CreateKind {
    Create,
    CreateIfNotExists,
    CreateOrReplace,
}

#[derive(Debug, Serialize)]
pub struct BoundDropGraphStatement {
    pub schema: SchemaRef,
    pub graph: Ident,
    pub if_exists: bool,
}

#[derive(Debug, Serialize)]
pub struct BoundCreateGraphTypeStatement {
    pub schema: SchemaRef,
    pub name: Ident,
    pub kind: CreateKind,
    pub source: BoundGraphType,
}

#[derive(Debug, Serialize)]
pub struct BoundDropGraphTypeStatement {
    pub schema: SchemaRef,
    pub graph_type: Ident,
    pub if_exists: bool,
}
