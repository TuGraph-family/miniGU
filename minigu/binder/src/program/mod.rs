use gql_parser::ast::{Procedure, Statement};
use minigu_catalog::provider::{CatalogRef, SchemaRef};

use crate::error::BindResult;
use crate::statement::procedure_spec::{BoundNextStatement, BoundProcedure, BoundStatement};
pub struct Binder {
    pub catalog: CatalogRef,
    pub schema: Option<SchemaRef>,
}

impl Binder {
    pub fn new(catalog_ref: CatalogRef, schema_ref: Option<SchemaRef>) -> Self {
        Binder {
            catalog: catalog_ref,
            schema: schema_ref,
        }
    }
}

impl Binder {
    pub fn bind_procedure(&mut self, procedure: &Procedure) -> BindResult<BoundProcedure> {
        if let Some(at_schema) = &procedure.at {
            let schema_ref = self.resolve_schema_ref(at_schema.value())?;
            self.schema = Some(schema_ref);
        }

        Ok(BoundProcedure {
            statement: self.bind_statement(procedure.statement.value())?,
            next_statement: procedure
                .next_statements
                .iter()
                .map(|stmt| {
                    Ok(BoundNextStatement {
                        yield_clause: None,
                        statement: self.bind_statement(stmt.value().statement.value())?,
                    })
                })
                .collect::<BindResult<Vec<_>>>()?,
        })
    }

    pub fn bind_statement(&mut self, statement: &Statement) -> BindResult<BoundStatement> {
        let mut resolved_statement = self.resolve_statement(statement)?;
        // self.type_check(&resolved_statement)?;
        // self.validate(&resolved_statement)?;
        Ok(resolved_statement)
    }
}
