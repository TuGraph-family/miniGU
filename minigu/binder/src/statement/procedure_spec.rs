use serde::Serialize;

use crate::object_ref::SchemaRef;
use crate::statement::catalog::LinearBoundCatalogModifyingStatement;

#[derive(Debug, Serialize)]
pub struct BoundProcedure {
    // pub binding_variable_def: BindingVariableDefBlock,
    pub statement: BoundStatement,
    pub next_statement: Vec<BoundNextStatement>,
}

#[derive(Debug, Serialize)]
pub enum BoundStatement {
    Catalog(LinearBoundCatalogModifyingStatement),
    // Query(BoundCompositeQueryStatement),
    // Data(BoundLinearDataModifyingStatement),
}

#[derive(Debug, Serialize)]
pub struct BoundNextStatement {
    pub yield_clause: Option<Vec<usize>>,
    pub statement: BoundStatement,
}
