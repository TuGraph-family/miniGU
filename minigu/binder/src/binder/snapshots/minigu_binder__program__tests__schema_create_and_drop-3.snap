---
source: minigu/binder/src/program/mod.rs
assertion_line: 229
expression: bound_drop_schema
---
Ok:
  at: ~
  binding_variable_def: []
  statement:
    Catalog:
      - DropSchema:
          path:
            segments:
              - a
              - b
              - c
          if_exists: true
  next_statement: []
